File: Python/Detect Capital.py
import string
class Solution(object):
    def detectCapitalUse(self, word):
        """
        :type word: str
        :rtype: bool
        """
        self.upper_set = set(string.ascii_uppercase)
        self.lower_set = set(string.ascii_lowercase)
        return self.case_one(word) or self.case_two(word) or self.case_three(word)
        
    def case_one(self, word):
        return set(word).issubset(self.upper_set)
    def case_two(self, word):
        return set(word).issubset(self.lower_set)
    def case_three(self, word):
        return word[0] in self.upper_set and set(word[1:]).issubset(self.lower_set)
        
    

File: cpp/Detect Capital.cpp
class Solution {
public:
    bool detectCapitalUse(string word) {
        unsigned int state = 0;
        for (char ch: word) {
            this->tran(state, ch);
            if (state == 5) {
                return false;
            }
        }
        return true;
    }
    void tran(unsigned int &state, char ch) {
        switch (state) {
            case 0: {
                if (ch >= 'a' && ch <= 'z') {
                    state = 1;
                } else {
                    state = 2;
                }
                break;
            }
            case 1: {
                if (ch >= 'A' && ch <= 'Z') {
                    state = 5;
                }
                break;
            }
            case 2: {
                if (ch >='a' && ch <= 'z') {
                    state = 3;
                } else {
                    state = 4;
                }
                break;
            }
            case 3: {
                if (ch >= 'A' && ch <= 'Z') {
                    state = 5;
                }
                break;
            }
            case 4: {
                if (ch >= 'a' && ch <= 'z') {
                    state = 5;
                }
                break;
            }
            default:
                state = 5;
        }
    }
};
